site/app/libraries/database/DatabaseQueries.php

```diff
@@ -2820,37 +2820,61 @@ public function getAllSectionsForGradeable($gradeable) {
+        $row_type = $user_type . "_id";
 
-        $this->course_db->query(
-            "
-             SELECT u.{$user_type}_id
-             FROM {$table} u
-                 JOIN (
-                     SELECT gr.sections_{$grade_type}_id
-                     FROM grading_{$grade_type} AS gr
-                     WHERE gr.user_id = ?
-                 ) AS gr
-                 ON gr.sections_{$grade_type}_id=u.{$grade_type}_section
-                 JOIN (
-                     SELECT gd.gd_{$type}_id, gcmd.gcm_id
-                     FROM gradeable_component_mark_data AS gcmd
-                         JOIN gradeable_data gd ON gd.gd_id=gcmd.gd_id
-                 ) as gcmd
-                 ON gcmd.gd_{$type}_id=u.{$type}_id
-             WHERE gcmd.gcm_id = ?",
-            $params
-        );
+        $params = [$grader->getId(), $mark->getId()];
+        $table = $mark->getComponent()->getGradeable()->isTeamAssignment() ? 'gradeable_teams' : 'users';
+        $grade_type = $gradeable->isGradeByRegistration() ? 'registration' : 'rotating';
+        if ($type == 'user' && $user_type == 'anon') {
+            array_unshift($params, $mark->getComponent()->getGradeable()->getId());
+            $this->course_db->query(
+                "
+                SELECT ga.{$user_type}_id
+                FROM {$table} u
+                    JOIN gradeable_anon ga ON ga.user_id=u.user_id AND ga.g_id = ?
+                    JOIN (
+                        SELECT gr.sections_{$grade_type}_id
+                        FROM grading_{$grade_type} AS gr
+                        WHERE gr.user_id = ?
+                    ) AS gr
+                    ON gr.sections_{$grade_type}_id=u.{$grade_type}_section
+                    JOIN (
+                        SELECT gd.gd_{$type}_id, gcmd.gcm_id
+                        FROM gradeable_component_mark_data AS gcmd
+                            JOIN gradeable_data gd ON gd.gd_id=gcmd.gd_id
+                    ) as gcmd
+                    ON gcmd.gd_{$type}_id=u.{$type}_id
+                WHERE gcmd.gcm_id = ?",
+                $params
+            );
+        }
+        else {
+            $this->course_db->query(
+                "
+                SELECT u.{$user_type}_id
+                FROM {$table} u
+                    JOIN (
+                        SELECT gr.sections_{$grade_type}_id
+                        FROM grading_{$grade_type} AS gr
+                        WHERE gr.user_id = ?
+                    ) AS gr
+                    ON gr.sections_{$grade_type}_id=u.{$grade_type}_section
+                    JOIN (
+                        SELECT gd.gd_{$type}_id, gcmd.gcm_id
+                        FROM gradeable_component_mark_data AS gcmd
+                            JOIN gradeable_data gd ON gd.gd_id=gcmd.gd_id
+                    ) as gcmd
+                    ON gcmd.gd_{$type}_id=u.{$type}_id
+                WHERE gcmd.gcm_id = ?",
+                $params
+            );
+        }
```

Comment:
Since these queries are quite similar, you could probably just concatenate strings to form this query to reduce the amount of duplicated code. There are a couple other queries in this file that do something like that.

Reply:
String concatenation is used now.